ui/app/ducks/swaps/swaps.js

```diff
@@ -165,7 +240,7 @@ export const getSwapsTradeTxParams = (state) => {
   }
   const { trade } = usedQuote
   const gas = getCustomSwapsGas(state) || trade.gas
-  const gasPrice = getCustomSwapsGasPrice(state) || trade.gasPrice
+  const gasPrice = getCustomSwapsGasPrice(state) || trade.gasPrice || getSwapsFallbackGasPrice(state)
```

Comment:
This is a pre-existing pattern, but wow I really don't like how this is being used :grimacing:.

I started tracking down where this was used, and noticed that the `quote.trade.gasPrice` property was being directly mutated in `signAndSendTransactions` with the `gasPrice` returned from this selector. But only if the `customGasPrice` wasn't already set (despite this _also_ being preferred by this selector directly).

But it's also used in various UI components (`view-quote.js`, `swaps/index.js`, `awaiting-swap.js`) without mutating the `quote.trade`object.

So I'm left wondering - _what is this selector doing_? Is this supposed to be the "authoritative" source of the "swaps trade tx params"? So `quote.trade` is _never_ to be used directly in a place where `txParams` are expected.

But if that's the case, then why does `signAndSendTransactions` build up its own `usedTradeTxParams` independently from this one? Is that a different thing? And why does `swaps/index.js` calculate `usedGasPrice` differently?

Reply:
I will clean this up

Reply:
@Gudahtt I believe that 9517b6a29 addresses your concerns

Most notably, it removes `getSwapsTradeTxParams`, as all cases of it being called could be replaced by either a new `const getUsedSwapsGasPrice = (state) => getCustomSwapsGasPrice(state) || getSwapsFallbackGasPrice(state)` or in the case of view-quote passing `tradeTxParams` to the swaps gas customization modal, `usedQuote` could be the source of the only txParam needed.

Also, given how the custom gas price is now set, and given that swaps submission is disabled if there is no gas price, there is never a need to default to `fastGasEstimate`. So that logic has been removed.

Reply:
I see - so `fastest` isn't used anymore because that should always be set as the "custom" price in the background, unless the API is down?

Reply:
correct

Reply:
Unfortunately I don't think it's guaranteed for the initial "custom" price to be set in every case where `getUsedSwapsGasPrice` is called. If it's unset, we may be using the fallback price here even in cases when the swaps gas API is up.

Reply:
@Gudahtt But the fallback price will only ever be set if we have already tried to get the initial gas price from the swaps api. So I think there are only three cases:
- the swaps api is up and custom price is set
- the swaps api is not up, and the fallback price is used
- either the swaps is not up AND we failed to fetch the fallback price OR the fetch for the swaps api is not complete / has not been made, in which case the price will be undefined

Reply:
>But the fallback price will only ever be set if we have already tried to get the initial gas price from the swaps api

This is the part that I'm doubting. I don't see how this is guaranteed at the moment - it looks like a race condition.

We fetch the initial gas price from the swaps API in a few different places, but none of them are blocking the use of this selector. For example, in `swaps/index.js` we fetch them as soon as the component mounts, but the render continues past there. This selector is used in places where it might be executed before it returns.

Reply:
Yes, you are right! This seems fine then.